### ISSUE TITLE
Blockchain/cryptocurrency proposal

### ISSUE THREAD

**Issue by alreadydone** [gcp/leela-zero#978]
Someone recently raised [on reddit/cbaduk](https://www.reddit.com/r/cbaduk/comments/80wtsg/i_am_doing_so_much_gpu_computation_for_leela_zero/) the idea to create a cryptocurrency with distributed game production and/or training. I've also thought about this before, and in the last two days my ideas seem to become mature, so I'm sharing them here, waiting for interested/skilled/experienced people/team to offer suggestions/comments and/or to implement them. People versed in probability/statistics (#543) and in the details of blockchains are invited to offer sensible estimates for the tunable parameters in this system, since validation is more probabilistic than in Bitcoin. Personally I would like to invest at least US$1,000 to get this project running (more if my confidence level rises from your feedback).

My proposal below is mostly modeled after Bitcoin, and encourages both the production of self-play games and of stronger networks:

### I. Self-play game production and rewards (mining)

+ Most people run lightweight clients to produce self-play games (or run matches) and receive, validate, cache, and send games (sgf + training data), blocks and transactions from other clients. A few people run "full nodes" which also store all blocks ever created.

+ Self-play games are produced with the latest version of leelaz, the current best network and a random seed generated by hashing the producer's address (public key) with some number (e.g. game count or timestamp), so no one can claim other people's work.

+ If the hash of a sgf lies in a certain range (depending on the latest block and an automatically adjusted "difficulty"), the producer has the right to create a new block containing all transactions and an index (including producer's addresses, timestamps, leelaz versions, network hashes, and game hashes) of all games received but not yet written into a block. The actual games are only stored by people who do the training as they are large in size. 

+ The creator of the new block receives a large reward (credit) and producers of the games inside the block receive a small reward for each game.

+ **The rewards are proportional to** (and only depend on) **the increase in Elo rating** (19x19, 7.5 komi) in the last few days. When 19x19 stalls the reward for creating a block will solely come from transaction fees. People must still produce many games to create a block though. 

+ One or multiple addresses can use their credit to post self-play jobs to be completed by other clients with customized settings (playout, board size, komi and handicap), rules (e.g. the [ancient Chinese rule](https://senseis.xmp.net/?GroupTax)), and/or variants ([3D](https://en.wikipedia.org/wiki/Diamond_cubic) [crystals](https://en.wikipedia.org/wiki/Lonsdaleite), [edgeless](https://senseis.xmp.net/?EdgelessGo) or [mosaic](https://tieba.baidu.com/f?kw=%E5%BC%82%E5%BD%A2%E7%89%88%E5%9B%BE%E5%9B%B4%E6%A3%8B&ie=utf-8) [Go](#753)) that they want to explore. They may even set up a bounty for a 100 Elo stronger network, etc., in the form of a smart contract. There could also be lottery for the outcome of block-creating games.

+ If Go enthusiasts own a large portion of credits when 19x19 stalls, they'll use them to post jobs; if miners/speculators own a large portion, they'll keep transactions active.

+ When 19x19 stalls the community could agree to devote (some of) the computing power to more worthy explorations with another setting/rule/variant, waiting for new technology to revive 19x19. A resource management algorithm could be programmed into the clients to do the same when training can't catch up with 19x19 self-play production rate. The reward scales with the current 19x19 reward / Elo increase and the difficulty of the job.

### II. Validation

+ Each client validates a received game by randomly sampling a fixed number of moves and check by running leelaz that the search probabilities in the training data are not too far off. The problematic moves will be broadcast and quickly verified independently.  (Tree reuse causes trouble here; we probably need to disable this, sacrificing some self-play strength, since storing the tree is rather costly.)

+ Each client validates a received block by randomly sampling a higher number of moves (to match the higher reward) from the block-creating game and check the search probabilities. The games and transactions in the block may also be checked (cf. Bitcoin).

+ A fraudulent game involving no work or substantially less work will be quickly rejected by the blockchain network. It's probably not worthy to produce a fraudulent game with just a bit less work and risk it being rejected. If one produces an authentic game by doing the required work, it makes no sense to edit the training data alone, but one may be tempted to edit the sgf (instead of playing another game) to get a desired hash for a higher reward, and edit the training data accordingly. However, to pass validation one could only change a few moves and must generate all remaining moves by doing actual work. By sampling more from the later moves, we can render such act unworthy.

+ Transactions are generated and validated the same way as in Bitcoin.

+ We could mimic the way the Bitcoin network deals with invalid games/blocks/transactions and nodes generating them.

### III. Miners and training

Professional miners with a lot of computing power (many fast GPUs) benefit the most from self-play production and block creation, so they have the strongest incentive to train stronger networks since Elo improvement increases their rewards, and because of the relatively small computing power required compared to self-play production. In addition to storing all blocks and serving as full nodes, they also store all games locally and have the training data at hand. Since they do not directly receive rewards from strong networks they train, they are willing to distribute authentic training data, since strong networks from other people also benefit them. If they prepare a lot of computing power before throwing in a significantly stronger network, they have the best chance to seize the booty, but this is beneficial to progress anyway. They also have incentives to

+ speed up progress by doing software development, optimizing self-play speed and strength, compressing network size, optimizing network architecture, adding in computationally efficient features (e.g. ladder reader, life-and-death module);
+ exploit more efficient training methods;
+ fairly reward self-play game producers (after some point the network can improve only from its own self-play games);
+ only throw in a reasonable amount of networks to avoid wasting to much computing power on matches.

Hopefully these miners will make @gcp "more redundant" as he [wished](#591).

Of course network size changes, client updates, and deviation from A(G)Z approach require the consensus of the community. I think most people just care about getting stronger and stronger networks, but if we want to stick to the A(G)Z approach we could also implement validation of networks: a node that throws in a new network need to provide specific information like: the training method (source code), which old network it was trained from, which blocks the training games came from, which random seed was used, and hyperparameters. It also needs to store a checkpoint every fixed number of steps, so that other people can sample some random step counts and verify their work independently before the new network is allowed for matches.

Matches need to be run within the blockchain network to get accurate Elo ratings for new networks that everyone agrees upon. I am thinking about a CGOS-like system: all weights are matched against each other to receive a Bayes Elo. Each node will use an algorithm to decide when to run a match between which two networks. A match receives the same reward as a self-play game.

------------
As for the name, I propose to call it WGB (Coin?) for Weiqi/Go/Baduk, pronounced as Wei-Go-Bee, for broader audience. You can also interpret as Whit-Go-Black if you wish. In Chinese, W could stand for Wei (围, surround), G for Go (棋/碁, board game), and B for Bi (币, currency), so WGB could be translated into 围棋币. LZC for Leela Zero Credit/Contribution/Coin is also good, since this project is where it originates from. Please post any suggestion that you consider better.
 
This model should be copiable to other board games and reinforcement learning projects, following the AI + blockchain trend. I have yet to see AlphaZero applied to Hex, Dots and Boxes, or Amazons.

Please help translate :) Chinese version to follow.

**Comment by alreadydone** [comment]
Proposal translated into other languages; to appear.

**Comment by ThorAvaTahr** [comment]
Haha,Cool idea! 
Not sure if I'm fully aboard, but just would like to give a first thumbs up.

**Comment by RavnaBergsndot** [comment]
kgs/tygem/fox high dan games are great sources for fabricating fake games. I think the move is similar enough to leela zero self play to convince the validator. Reducing the playout number from 1600 to, say 200, may also get away with the validator.

**Comment by kris-computer-go** [comment]
I think we are interesting here in GO game and NOT cryptocurrency of any kind.

**Comment by alreadydone** [comment]
@RavnaBergsndot I would like validation to verify the search probabilities (aka visit counts, contained in training data) for each move sampled, that is 362 floating point numbers (of course some error margin is allowed). I don't think 200 playouts will get close to the results of 1600 playouts; for example it will leave many more moves unvisited.

**Comment by alreadydone** [comment]
@kris-computer-go The idea is to speed up progress by summoning computing power.

**Comment by RavnaBergsndot** [comment]
The search possibilities of unvisited moves in both 200 playout fake games and kgs based fake games (which have only one visited node per move) can be fabricated using the some simple statistics. The real 1600 playout games have noise in the root node anyway so any validation methods would have a large error margin.

**Comment by alreadydone** [comment]
@RavnaBergsndot The noise should be completely determined by the random seed as it currently is, but games aren't completely reproducible now because of differences between GPUs/drivers. I seem to have read that games are reproducible on the same kind of GPUs. I agree that we should do some tests to find out how easy it is to fabricate games and how to set the error margin.

**Comment by RavnaBergsndot** [comment]
Even if the noise can be determined by the random seed, the validation should not be just run 1600 playouts on a random move. It should be much faster than that.

If the validation is full validating a move per game (with 250 moves in average), the validation process would be 1/250 as hard as the whole network's game producing overload. It's like each Bitcoin node has to do 1/250 of the mining of the whole Bitcoin network just for validation.

If the validation can be much faster than that, then it should be statistical one, not a replication-based one. It means we can't make use of the random seed.

**Comment by jkiliani** [comment]
This is somewhat related to the BOINC proposal, where there is also planned to be proof of work (with some amount of duplication involved) and a distribution of tokens as a reward.

I see the principal difficulty in such a plan in that it creates an incentive for people to cheat the system by feeding junk data to the project. Right now, pure vandalism is the only possible reason for people to hack clients and feed garbage data to the server, but with a financial reward involved that would no longer be the case. I would be very surprised if this would not result in quite a number of attempts to siphon off the reward pool (crypto coins etc.) with easy to produce bad data. As long as enough legit clients participate in the effort, improvement may still happen, but if the bad clients manage to get around the safeguards, they will be credited with that improvement instead of the clients producing the real self-play games.

Maybe there are ways to validate training data that is resistant to such attempts, but from my understanding of the current state of game reproducibility, it would be very difficult to implement such safeguards.

**Comment by RavnaBergsndot** [comment]
Yeah, getting blockchain PoW to do useful work is a hard problem in general. There's a reason why we still don't have any meaningful physics emulation, protein folding, alien finding, or whatever distributed projects successfully run on a blockchain yet, despite all the incentives and academic motivations.

**Comment by gcp** [comment]
Discussion about cryptocurrencies is not welcome here.

**Comment by alreadydone** [comment]
A new proposal is out, please see: https://www.reddit.com/r/cbaduk/comments/888jy2/leela_zero_credits_a_cryptocurrency_proposal/
